log4j-core/src/main/java/org/apache/logging/log4j/core/LoggerContext.java

```diff
@@ -92,6 +94,14 @@ public class LoggerContext extends AbstractLifeCycle
     private volatile URI configLocation;
     private Cancellable shutdownCallback;
 
+    /**
+     * The default message factory to use while creating loggers.
+     */
+    private final Lazy<MessageFactory> defaultMessageFactoryRef = Lazy.lazy(() -> {
+        final Logger throwawayLogger = newInstance(this, "throwaway-for-determining-MF", null);
+        return throwawayLogger.getMessageFactory();
+    });
```

Comment:
Using this would mean that `newInstance` is called **once** with a `null` message factory and the message factory of that bogus `Logger` is used in every other `newInstance` invocation. 
So instead of allowing the user to determine the default message factory directly, we allow him to do it indirectly by placing some special logic in `newInstance`.

If you really believe that users want to change the default message factory programmatically, let's add a protected `getMessageFactory(String)` method.

Personally I don't think this is required. A default message factory must work exactly like `ParameterizedMessageFactory`, only the implementation details can change, but `getFormattedMessage()` of the messages it produces must be identical to `PMF`.

Reply:
Fixed in c1dff86c087c9e4ae5ee7e68ad16385d02dc894e as you wanted. (For the record, I don't agree with this direction, but there are more pressing issues than discussing this detail.)